File: src/main/java/com/ptaf/utils/BrowserFactory.java
package com.ptaf.utils;

import com.microsoft.playwright.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.file.Paths;

/**
 * BrowserFactory is a utility class for creating Playwright Browser instances.
 * It abstracts the browser creation logic, allowing users to easily instantiate
 * different types of browsers (Chrome, Firefox, WebKit, Microsoft Edge) based on their requirements.
 */
public class BrowserFactory {

    private static final Logger logger = LoggerFactory.getLogger(BrowserFactory.class);
    static String headlessMode = ConfigurationProperties.getHeadlessMode();
    static String videoCapture = ConfigurationProperties.getVideoCapture(); // New property
    private static final String VIDEO_DIR = "test-output/captured-videos";


    /**
     * Enum representing the supported browser types.
     */
    public enum BrowserTypeEnum {
        CHROME,
        FIREFOX,
        WEBKIT,
        EDGE
    }

    /**
     * Creates and launches a Playwright Browser instance based on the specified browser type.
     *
     * @param browserTypeEnum The type of browser to create.
     * @return A Playwright Browser instance.
     */
    public static Browser createBrowser(BrowserTypeEnum browserTypeEnum) {
        Playwright playwright = Playwright.create();
        return switch (browserTypeEnum) {
            case CHROME -> {
                BrowserType browserType = playwright.chromium();
                yield launchBrowser(browserType);
            }
            case FIREFOX -> {
                BrowserType browserType = playwright.firefox();
                yield launchBrowser(browserType);
            }
            case WEBKIT -> {
                BrowserType browserType = playwright.webkit();
                yield launchBrowser(browserType);
            }
            case EDGE -> {
                boolean headless = Boolean.parseBoolean(headlessMode);
                logger.info("Launching Microsoft Edge with headless mode: {}", headless);
                yield playwright.chromium().launch(new BrowserType.LaunchOptions()
                        .setChannel("msedge")
                        .setHeadless(headless));
            }
        };
    }

    /**
     * Launches the specified browser type with the configured headless mode.
     *
     * @param browserType The BrowserType instance.
     * @return The launched Browser.
     */
    private static Browser launchBrowser(BrowserType browserType) {
        boolean headless = Boolean.parseBoolean(headlessMode);
        logger.info("Launching browser: {} with headless mode: {}", browserType.name().toUpperCase(), headless);
        return browserType.launch(new BrowserType.LaunchOptions().setHeadless(headless));
    }

    /**
     * Creates a browser context with optional video capture.
     *
     * @param browser The Browser instance.
     * @return A BrowserContext with or without video enabled.
     */
    public static BrowserContext createContextWithVideo(Browser browser) {
        boolean recordVideo = Boolean.parseBoolean(videoCapture);
        Browser.NewContextOptions contextOptions = new Browser.NewContextOptions();

        if (recordVideo) {
            logger.info("Video capture enabled.");
            contextOptions.setRecordVideoDir(Paths.get(VIDEO_DIR))
                    .setRecordVideoSize(1280, 720);
        } else {
            logger.info("Video capture disabled.");
        }

        return browser.newContext(contextOptions);
    }
}

File: ptaf/utils/browser_factory.py
"""Playwright browser and context factory (ported from BrowserFactory.java)."""

from __future__ import annotations

import logging
from enum import Enum
from pathlib import Path

from playwright.sync_api import Browser, BrowserContext, Playwright

from ptaf.utils import config

logger = logging.getLogger(__name__)

VIDEO_DIR = Path(__file__).resolve().parent.parent.parent / "test-output" / "captured-videos"


class BrowserTypeEnum(str, Enum):
    CHROME = "CHROME"
    FIREFOX = "FIREFOX"
    WEBKIT = "WEBKIT"
    EDGE = "EDGE"


def parse_browser_type(browser_name: str) -> BrowserTypeEnum:
    normalized = browser_name.strip().upper()
    try:
        return BrowserTypeEnum(normalized)
    except ValueError as exc:
        raise ValueError(f"Unsupported browser type: {browser_name}") from exc


def create_browser(playwright: Playwright, browser_type: BrowserTypeEnum) -> Browser:
    headless_raw = config.get_headless_mode() or "false"
    headless = headless_raw.strip().lower() == "true"

    match browser_type:
        case BrowserTypeEnum.CHROME:
            logger.info(
                "Launching browser: CHROMIUM with headless mode: %s", headless
            )
            return playwright.chromium.launch(headless=headless)
        case BrowserTypeEnum.FIREFOX:
            logger.info(
                "Launching browser: FIREFOX with headless mode: %s", headless
            )
            return playwright.firefox.launch(headless=headless)
        case BrowserTypeEnum.WEBKIT:
            logger.info(
                "Launching browser: WEBKIT with headless mode: %s", headless
            )
            return playwright.webkit.launch(headless=headless)
        case BrowserTypeEnum.EDGE:
            logger.info("Launching Microsoft Edge with headless mode: %s", headless)
            return playwright.chromium.launch(channel="msedge", headless=headless)


def create_context_with_video(browser: Browser) -> BrowserContext:
    video_capture_raw = config.get_video_capture() or "false"
    record_video = video_capture_raw.strip().lower() == "true"

    context_options: dict[str, object] = {"ignore_https_errors": True}

    if record_video:
        logger.info("Video capture enabled.")
        VIDEO_DIR.mkdir(parents=True, exist_ok=True)
        context_options["record_video_dir"] = str(VIDEO_DIR)
        context_options["record_video_size"] = {"width": 1280, "height": 720}
    else:
        logger.info("Video capture disabled.")

    return browser.new_context(**context_options)
